cupy/random/_bit_generator.pyx

```diff
@@ -0,0 +1,184 @@
+    cppclass curandStateMRG32k3a:
+        pass
+    cppclass curandStatePhilox4_32_10_t:
+        pass
+
+    cdef enum _RandGenerators '::RandGenerators':
+        CURAND_XOR_WOW
+        CURAND_MRG32k3a
+        CURAND_PHILOX_4x32_10
+
+    void init_curand_generator(int generator, intptr_t state_ptr, uint64_t seed, ssize_t size, intptr_t stream);
+    void interval_32(int generator, intptr_t state, intptr_t out, ssize_t size, intptr_t stream, int mx, int mask);
+    void interval_64(int generator, intptr_t state, intptr_t out, ssize_t size, intptr_t stream, uint64_t mx, uint64_t mask);
+    void beta(int generator, intptr_t state, intptr_t out, ssize_t size, intptr_t stream, double a, double b);
+    void exponential(int generator, intptr_t state, intptr_t out, ssize_t size, intptr_t stream);
+
+_UINT32_MAX = 0xffffffff
+_UINT64_MAX = 0xffffffffffffffff
+
+
+class BitGenerator:
+    def __init__(self, seed=None):
+        self.lock = threading.Lock()
+        # If None, then fresh, unpredictable entropy will be pulled from the OS.
+        # If an int or array_like[ints] is passed, then it will be passed 
+        # to ~`numpy.random.SeedSequence` to derive the initial BitGenerator state.
+        # TODO(ecastill) port SeedSequence
+        self._seed_seq = numpy.random.SeedSequence(seed)
+        dev = cupy.cuda.Device()
+        self._current_device = dev.id
+
+    def random_raw(self, size=None, out=False):
+        raise NotImplementedError(
+            'Subclasses of `BitGenerator` must override `random_raw`')
+
+    def state_size(self):
+        """Maximum number of samples that can be generated at once
+        """
+        return 0
+
+    def _check_device(self):
+        if cupy.cuda.Device().id != self._current_device:
+            raise RuntimeError("This Generator state has been allocated in a different device")
+
+
+class _cuRANDGenerator(BitGenerator):
+    # Size is the number of threads that will be initialized
+    def __init__(self, seed=None, size=1024*100):
+        super().__init__(seed)
+        self._seed = self._seed_seq.generate_state(1, numpy.uint64)[0]
+        self._size = size
+        cdef ssize_t b_size = self._type_size() * size
+        self._state = cupy.zeros(b_size, dtype=numpy.int8)
+        ptr = self._state.data.ptr
+        cdef intptr_t state_ptr = <intptr_t>ptr
+        cdef uint64_t c_seed = <uint64_t>self._seed
+        cdef intptr_t _strm = stream.get_current_stream_ptr()
+        # Initialize the state
+        init_curand_generator(self.generator, state_ptr, self._seed, size, _strm)
+
+    def random_raw(self, size=None, out=False):
+        pass
+
+    def state(self):
+        self._check_device()
+        return self._state.data.ptr
+
+    def state_size(self):
+        return self._size
+
+    def _type_size(self):
+        return 0 
+
+class XORWOW(_cuRANDGenerator):
+    generator = CURAND_XOR_WOW  # Use The Enum
+
+    def _type_size(self):
+        return sizeof(curandState)
```

Comment:
I am a bit confused: what's the difference between `curandState` and `curandState_t`?

Reply:
no one knows, they are used interchangeably in the documentation, but we should stick to only one, will fix